### PR TITLE
[SPARK-28695][SS] Use CaseInsensitiveMap in KafkaSourceProvider to make source param handling more robust

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
## What changes were proposed in this pull request?

[SPARK-28163](https://issues.apache.org/jira/browse/SPARK-28163) fixed a bug and during the analysis we've concluded it would be more robust to use `CaseInsensitiveMap` inside Kafka source. This case less lower/upper case problem would rise in the future.

Please note this PR doesn't intend to solve any kind of actual problem but finish the concept added in [SPARK-28163](https://issues.apache.org/jira/browse/SPARK-28163) (in a fix PR I didn't want to add too invasive changes). In this PR I've changed `Map[String, String]` to `CaseInsensitiveMap[String]` to enforce the usage. These are the main use-cases:
* `contains` => `CaseInsensitiveMap` solves it
* `get...` => `CaseInsensitiveMap` solves it
* `filter` => keys must be converted to lowercase because there is no guarantee that the incoming map has such key set
* `find` => keys must be converted to lowercase because there is no guarantee that the incoming map has such key set
* passing parameters to Kafka consumer/producer => keys must be converted to lowercase because there is no guarantee that the incoming map has such key set

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
cc @HeartSaVioR @HyukjinKwon @cloud-fan this is a follow-up on https://github.com/apache/spark/pull/24967

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala [hunk 1]
```diff
@@ -307,14 +317,14 @@ private[kafka010] class KafkaSourceProvider extends DataSourceRegister
...
-  private def validateStreamOptions(caseInsensitiveParams: Map[String, String]) = {
+  private def validateStreamOptions(caseInsensitiveParams: CaseInsensitiveMap[String]) = {
     // Stream specific options
     caseInsensitiveParams.get(ENDING_OFFSETS_OPTION_KEY).map(_ =>
       throw new IllegalArgumentException("ending offset not valid in streaming queries"))
     validateGeneralOptions(caseInsensitiveParams)
   }
 
-  private def validateBatchOptions(caseInsensitiveParams: Map[String, String]) = {
+  private def validateBatchOptions(caseInsensitiveParams: CaseInsensitiveMap[String]) = {
```
Comment (cloud-fan) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: maybe we can rename the parameter name to `params`, as we can easily tell it's case insensitive by the type.

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: Good simplification, on the way...

Comment (gaborgsomogyi) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceProvider.scala]: Renamed on the places where it's not used together with the non case sensitive one + fixed some style problems.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**